### PR TITLE
build: Speed up docker build in some setups

### PR DESCRIPTION
**Pull request by abador**
This change adds builder cache in dockerfile. This speeds up the build in some cases.
## Checklist

- [x ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x ] I have read the [security policy](../security/policy).
- [ x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x ] I have added tests that prove my fix is effective or that my feature
      works.
- [x ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
Could you please link some docs explaining how this works? Thank you :)

**Comment by abador** [comment]
This change is basically a way to add caching to our docker builds on a local machine that runs that build. If you run it on a different machine there is no difference. I also split this into a multi-stage build (this also might speed some things up). Some example doc about this: https://www.docker.com/blog/containerize-your-go-developer-environment-part-2/

File: .docker/Dockerfile-build [hunk 1]
```diff
@@ -1,4 +1,5 @@
-FROM golang:1.16-alpine AS builder
+# syntax = docker/dockerfile:1-experimental
```
Comment (aeneasr) [inline .docker/Dockerfile-build:1]: Is this needed?

File: .docker/Dockerfile-build [hunk 2]
```diff
@@ -13,13 +14,15 @@ ENV CGO_ENABLED 1
 
 RUN go mod download
 
+FROM base as builder
```
Comment (aeneasr) [inline .docker/Dockerfile-build:17]: We are using the same image as in the previous step, do we need to or can we leave this out?

Comment (abador) [reply .docker/Dockerfile-build:17]: it will still work porperly without this

File: .docker/Dockerfile-build (hunk 1 above)
Comment (abador) [reply .docker/Dockerfile-build:1]: As far as I know this changes the way dockerfiles are parsed and it's still needed in the current version

**Pull request closed without merging**